cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/join/Join.java

```diff
@@ -73,31 +80,34 @@ public void joinRecords() throws Exception {
     return perStageJoinElements;
   }
 
-  private void innerJoin(Map<String, List<JoinElement<INPUT_RECORD>>> perStageJoinElements) throws Exception {
-    // As we get intersection of records from n stages in inner join, if the number of stages we got after reduce
-    // are not same as number of stages we are joining, then there is nothing to emit.
-    if (perStageJoinElements.size() != numOfInputs) {
-      return;
-    }
-
+  private void join(Map<String, List<JoinElement<INPUT_RECORD>>> perStageJoinElements, Set<String> requiredInputs)
+    throws Exception {
     List<List<JoinElement<INPUT_RECORD>>> list = new ArrayList<>(perStageJoinElements.values());
     ArrayList<JoinElement<INPUT_RECORD>> joinElements = new ArrayList<>();
-    getCartesianProduct(list, 0, numOfInputs, joinElements);
+    Set<String> includedInputs = new HashSet<>();
+    getCartesianProduct(list, 0, joinElements, includedInputs, requiredInputs);
   }
 
   // TODO use iterative algorithm instead of recursion
-  private void getCartesianProduct(List<List<JoinElement<INPUT_RECORD>>> list, int index, int size,
-                                   List<JoinElement<INPUT_RECORD>> joinElements) throws Exception {
-    if (joinElements.size() == size) {
+  private void getCartesianProduct(List<List<JoinElement<INPUT_RECORD>>> list, int index,
+                                   List<JoinElement<INPUT_RECORD>> joinElements,
+                                   Set<String> includedInputs, Set<String> requiredInputs) throws Exception {
+    // check till the end of the list and emit only if records from all the required inputs are present in joinElements
+    if (index == list.size() && includedInputs.equals(requiredInputs)) {
```

Comment:
the name 'includedInputs' is a little misleading.  You can probably change this to includedInputs.containsAll(requiredInputs) and remove the if condition on line 103.